### PR TITLE
Use and extend view components in Solidus Admin

### PR DESCRIPTION
**Pull request by waiting-for-dev**
## Summary

We're introducing [ViewComponent](https://viewcomponent.org/) for the new admin along with proposing a way to extend them from the host app.

Please, go through each commit description for a more complete walk-through.

As a recap, admin components can be replaced by the host app by creating a new component with a matching path within the `app/components/{app_name}/solidus_admin/` directory. Example:

```ruby
# app/components/sandbox/solidus_admin/main_nav_item_component.rb

module Sandbox
  module SolidusAdmin
    class MainNavItemComponent < ViewComponent::Base
      with_collection_parameter :item
  
      attr_reader :item
  
      def initialize(item:)
        @item = item
      end
  
      erb_template <<~ERB
        <li style="background-color: red">
          <a href="#">
            <%= item.title %>
          </a>
        </li
      ERB
    end
  end
end
```

In the case of desired fine-grained control, the created component can inherit the original one. For instance, here we're just changing the template while keeping all the rest of view logic in place (and the opposite could also be possible):

```ruby
# app/components/sandbox/solidus_admin/main_nav_item_component.rb

module Sandbox
  module SolidusAdmin
    class MainNavItemComponent < ::SolidusAdmin::MainNavItemComponent
      erb_template <<~ERB
        <li style="background-color: red">
          <a href="#">
            <%= item.title %>
          </a>
        </li
      ERB
    end
  end
end
```

Unfortunately, view_component[ doesn't allow overriding templates](https://github.com/ViewComponent/view_component/issues/411) from the host application by just having the same relative path. For now, we'll use the [inline templates](https://viewcomponent.org/guide/templates.html#inline) feature

Worth noting that we're injecting nested components as initialization dependencies for their parent components instead of going with slots. That's to avoid the problem with [deeply nested slots](https://github.com/ViewComponent/view_component/discussions/867) and also to make customization easier by just overriding the initializer if needed:

```ruby
# app/components/sandbox/solidus_admin/main_nav_component.rb

module Sandbox
  module SolidusAdmin
    class MainNavComponent < ::SolidusAdmin::MainNavComponent
      def initialize(main_nav_item_component: component("my_main_nav_item"), **kwargs)
        super
      end
    end
  end
end
```
``
The same registry that is used to keep track of the components is also used to contain data entities, like the one used to wrap the main navigation items. For instance, a new nav item could be registered like this:

```ruby
require "solidus_admin/nav_item"
stats = SolidusAdmin::Stats.new(title: "Stats", position: 15)
SolidusAdmin::Container.register("main_nav.stats", stats)
```

Not showcased now, but we can also use the registry for the nav item itself, so users would have also full control over them.

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

### CODE REVIEW

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (tvdeyen) [inline admin/app/components/solidus_admin/main_nav_component.rb]: What do you think about?

```rb
components.all("main_nav")...
```

as helper for the `container` implementation detail?

File: admin/app/components/solidus_admin/base_component.rb
Comment (tvdeyen) [inline admin/app/components/solidus_admin/base_component.rb:7]: What about `SolidusAdmin::ComponentsHelper`? I would like to avoid adding a technical implementation detail to our public API. I prefer naming that explains its usecase rather than its implementation.

File: admin/app/helpers/solidus_admin/container_helper.rb
Comment (tvdeyen) [inline admin/app/helpers/solidus_admin/container_helper.rb:11]: 👏🏻 I like this as this hides the implementation.

File: admin/lib/solidus_admin/system/providers/main_nav.rb
Comment (tvdeyen) [inline admin/lib/solidus_admin/system/providers/main_nav.rb:7]: Maybe a `Components` namespace would help here as well to understand the usecase of this "Provider" a bit better?

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (tvdeyen) [inline admin/app/components/solidus_admin/main_nav_component.rb:6]: If we would add a `components` helper that uses `container.within_namespace` internally, we could make this much more userfriendly and would hide the technical detail of a `Dry::Container` from the user.

```suggestion
    def initialize(main_nav_item_component: component("main_nav_item"), items: components("main_nav"))
```

WDYT?

**Review by tvdeyen — COMMENTED** [review]
Yeah `view_component` 🎉 

I really like that we are going this route 🥳 

First of all I want to emphasize the great writing and git work on this PR. This is how PRs and commits should be written. It has been a pleasure to review and future version of ourslevs will thank you for this. 🎖️ 

I have some thoughts on naming.

File: admin/lib/solidus_admin/container.rb
Comment (waiting-for-dev) [reply admin/lib/solidus_admin/container.rb:16]: The issue is that, probably, we'll use the container for registering other kinds of dependencies. For instance, if, at some point, we introduce service objects only for the admin panel, they would also need to be registered here. In fact, we're already using it for an entity-like structure: the data wrapped within a `SolidusAdmin::MainNavItem` shouldn't be seen as something that only belongs to components.

File: admin/lib/solidus_admin/main_nav_item.rb
Comment (waiting-for-dev) [reply admin/lib/solidus_admin/main_nav_item.rb:5]: As pointed in https://github.com/solidusio/solidus/pull/5099#discussion_r1207065776, a `MainNavItem` is just something similar to a data object. We probably don't want to see it as something that can only belong to a view component. For instance, if at some point we dedice to persist them, they could become database records.

File: admin/app/components/solidus_admin/main_nav_component/main_nav_component.html.erb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/main_nav_component/main_nav_component.html.erb]: Fixed in subsequent commits :slightly_smiling_face:

File: admin/app/helpers/solidus_admin/container_helper.rb
Comment (waiting-for-dev) [reply admin/app/helpers/solidus_admin/container_helper.rb:8]: I also have some doubts about the name, what do you think about `SolidusAdmin::Registry`?

File: admin/app/components/solidus_admin/base_component.rb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/base_component.rb]: See https://github.com/solidusio/solidus/pull/5099#discussion_r1207108938

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/main_nav_component.rb]: See https://github.com/solidusio/solidus/pull/5099#discussion_r1207108938

File: admin/app/components/solidus_admin/base_component.rb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/base_component.rb:7]: See https://github.com/solidusio/solidus/pull/5099#discussion_r1207108938

File: admin/lib/solidus_admin/system/providers/main_nav.rb
Comment (waiting-for-dev) [reply admin/lib/solidus_admin/system/providers/main_nav.rb:7]: See https://github.com/solidusio/solidus/pull/5099#discussion_r1207065776

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/main_nav_component.rb:6]: Please, notice that `container.within_namespace("main_nav")` is not resolving `ViewContainer` classes nor instances, but instances of `SolidusAdmin::MainNavItem` entities. I don't think it's a big problem exposing the container (happy to rename it to registry) to the components, as it's precisely intended to resolve dependencies. However, I agree that there's a problem having it also available in the templates, but I can't see a way to avoid that, as it's more of a `view_component` limitation. What do you think?

On top of that, I'd love to use a more declarative specification of dependencies via [`dry-system`'s auto-injection](https://dry-rb.org/gems/dry-system/1.0/dependency-auto-injection/) feature. However, that would require having to call `super` everytime we define the `initialize` method for a component. I didn't go that way because I think it's good to stick as much as we can to bare `view_components` so that ours users don't have a hard time extending our components.

```ruby
module SolidusAdmin
  class MainNavComponent < BaseComponent
    include Import["main_nav_item_component"]

    def initialize(items: container.within_namespace("main_nav"))
      super
      @items = items
    end

    erb_template <<~ERB
      <nav>
        <%=
          render main_nav_item_component.with_collection(
            sorted_items
          )
        %>
      </nav>
    ERB

    private

    def sorted_items
      @items.sort_by(&:position)
    end
  end
end
```

**Comment by waiting-for-dev** [comment]
> I really like that we are going this route partying_face

Happy to hear that! I hope our broader community will feel the same :slightly_smiling_face: 

> First of all I want to emphasize the great writing and git work on this PR. This is how PRs and commits should be written. It has been a pleasure to review and future versions of ourselves will thank you for this. medal_military

Thanks for your nice words :bow: :slightly_smiling_face:

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (kennyadsl) [inline admin/app/components/solidus_admin/main_nav_component.rb:12]: Is tailwind able to catch the classes of these HTML elements even if they are in the .rb file?

File: admin/lib/solidus_admin/container.rb
Comment (kennyadsl) [inline admin/lib/solidus_admin/container.rb:5]: I'm not against adding this dependency, but just to be aware of all the pieces of the tradeoff we are doing here, do you have an idea of the effort needed to create something similar internally? I have the feeling it's not a giant effort, and maybe it might make sense only if we use small parts of what's provided in dry container/component; interested in your take on this.

File: admin/lib/solidus_admin/system/providers/main_nav.rb
Comment (kennyadsl) [inline admin/lib/solidus_admin/system/providers/main_nav.rb:11]: Do you envision any problem with when we'll need to use I18n here for the title?

**Review by kennyadsl — COMMENTED** [review]
Thanks @waiting-for-dev, I left some questions before approving, but I'm overall excited to merge this!

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/main_nav_component.rb:12]: We should be fine if we add the `components` directory to the Tailwind configuration. [Tailwind uses ](https://tailwindcss.com/docs/content-configuration#class-detection-in-depth)a very simple regexp matching to extract the needed classes.

File: admin/lib/solidus_admin/system/providers/main_nav.rb
Comment (waiting-for-dev) [reply admin/lib/solidus_admin/system/providers/main_nav.rb:11]: I don't think so, as `I18n` acts globally. Alternatively, we could use the I18n key and translate from the component.

File: admin/lib/solidus_admin/container.rb
Comment (kennyadsl) [reply admin/lib/solidus_admin/container.rb:5]: Makes sense, let's re-evaluate if we really need it once we are ok with the architecture and we see it in action. Thanks!

**Review by kennyadsl — APPROVED** [review]
Approving, but there's an issue with the installer related to the version of the `solidus_admin`. Do we already have a fix for that in mind?

**Comment by waiting-for-dev** [comment]
> Approving, but there's an issue with the installer related to the version of the `solidus_admin`. Do we already have a fix for that in mind?

Not sure what's going on. I tried to reproduce locally but it worked. I'll look into it.

**Comment by waiting-for-dev** [comment]
> Approving, but there's an issue with the installer related to the version of the `solidus_admin`. Do we already have a fix for that in mind?

@kennyadsl, bumping the cache version on the CI to fetch the latest changes on SSF to allow `view_components ~> 3.0` fixed the issue (ref: https://github.com/solidusio/solidus_starter_frontend/pull/343)

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/main_nav_component.rb:12]: FYI, I rebased from the admin branch, which now contains tailwindcss, and pushed a new commit adding the component directories to the list of scanned paths.

Comment (kennyadsl) [reply admin/app/components/solidus_admin/main_nav_component.rb:12]: 💯

**Review by elia — COMMENTED** [review]
It's great to see this coming along!

Left a couple of comments, mainly food for thought, happy to discuss it with a shared screen!

File: admin/lib/solidus_admin/container.rb
Comment (waiting-for-dev) [reply admin/lib/solidus_admin/container.rb:23]: Hey @elia, as pointed out in a [previous comment](https://github.com/solidusio/solidus/pull/5099#discussion_r1210069301), there's definitely nothing we couldn't build on our own. However, for the sake of this spike, I wanted to focus on the architecture and the interface to interact with it instead of getting lost in implementation details. dry-system allowed me to do that, but I have no problems revisiting it and evaluating if it's enough to build something like a Hash-based registry.

I'm not sure which direction we're going to take on other topics, like service objects on the new admin, but having dry-system in place can help us with moving and changing things faster, as its declarative configuration is very flexible. Then, it can also give us other valuable features like logging, monitoring, or testing support (stubbing a component while testing). We can assess how much of it we end up using, and then decide whether to go with or without it. What do you think?

File: admin/app/views/layouts/solidus_admin/application.html.erb
Comment (waiting-for-dev) [reply admin/app/views/layouts/solidus_admin/application.html.erb:8]: Hey @elia, I pointed out some of the befits and features that dry-system provides in [another answer](https://github.com/solidusio/solidus/pull/5099#discussion_r1217434038).

> <%= render component("main_nav").new(items: SolidusAdmin::NavItem.all) %>

I think it's good having the default value for `items` already defined on the initializer of `MainNavComponent`. In this case, it doesn't make such a difference, but maybe other components will be reused, which would avoid boilerplate code. In that situation, dry-system's [auto_inject](https://dry-rb.org/gems/dry-system/1.0/dependency-auto-injection/) would also help.

> class SolidusAdmin::NavItem < ActiveModel::Model
>  def self.all
>    @all ||= SolidusAdmin::Config.nav_items.map { new _1 }
>  end
> end

I know that's not a formal proposal, but regardless of using dry-system or not, I'd go with a PORO for nav item entities. In the end, we don't need to deal with validations, which would be a case for `ActiveModel::Model`.

> SolidusAdmin::Config.nav_items += [
>  { name: "Extra item", icon: :gift, … }
> ]

I'd also like to avoid having `SolidusAdmin::Config` as a hodgepodge. For anything else that configuring simple primitives, I think it'd be good having a richer interface, whether that one is provided by dry-system or something we own.

Comment (waiting-for-dev) [reply admin/app/views/layouts/solidus_admin/application.html.erb:8]: @elia, please, take a look at the last commit I pushed (__Auto-inject dependencies into components__). I also included [`dry-sytem`'s auto_inject](https://dry-rb.org/gems/dry-system/1.0/dependency-auto-injection/) within this spike, to better showcase both how having defaults defined in the `#initialize` method, and having nav items registered in the container help with avoiding boilerplate. Also pinging @kennyadsl.

File: admin/app/components/solidus_admin/main_nav_component.rb
Comment (waiting-for-dev) [reply admin/app/components/solidus_admin/main_nav_component.rb]: Finally, I also added auto-injection to this spike. In the end, `super` needs to be called whenever you override an initializer, and rubocop will complain if you don't.

File: admin/lib/solidus_admin/container.rb
Comment (elia) [reply admin/lib/solidus_admin/container.rb:23]: Maybe it's unfamiliarity with DRY, but looks like for our case it's an additional layer of injections on top of the already sufficient initializer params, which I also find more explicit.

I think having a more complete context might help in appreciating pros and cons of each solution, thoughts on holding a little bit on merging and keeping this PR as a reference maybe coming back to it once we start implementing? 

_I think we can merge right away the first commits setting up the Base component class_

Comment (waiting-for-dev) [reply admin/lib/solidus_admin/container.rb:23]: After talking offline, we decided to merge everything now and re-evaluate along the way :+1:

**Pull request merged**
